crates/provider-wadm/src/lib.rs

```diff
@@ -0,0 +1,583 @@
+        handlers.clear();
+
+        // clear the consumer components
+        let mut consumers = self.consumer_components.write().await;
+        consumers.clear();
+
+        // dropping all connections should send unsubscribes and close the connections, so no need
+        // to handle that here
+        Ok(())
+    }
+}
+
+impl bindings::exports::wasmcloud::wadm::client::Handler<Option<Context>> for WadmProvider {
+    #[instrument(level = "debug", skip(self, ctx), fields(model_name = %model_name))]
+    async fn deploy_model(
+        &self,
+        ctx: Option<Context>,
+        model_name: String,
+        version: Option<String>,
+        lattice: Option<String>,
+    ) -> anyhow::Result<Result<String, String>> {
+        let client = self.get_client(ctx).await?;
+        match client
+            .deploy_manifest(&model_name, version.as_deref())
+            .await
+        {
+            Ok((name, _version)) => Ok(Ok(name)),
+            Err(err) => {
+                error!("Deployment failed: {err}");
+                Ok(Err(format!("Deployment failed: {err}")))
+            }
+        }
+    }
+
+    #[instrument(level = "debug", skip(self, ctx), fields(model_name = %model_name))]
+    async fn undeploy_model(
+        &self,
+        ctx: Option<Context>,
+        model_name: String,
+        lattice: Option<String>,
+        non_destructive: bool,
+    ) -> anyhow::Result<Result<(), String>> {
+        let client = self.get_client(ctx).await?;
+        match client.undeploy_manifest(&model_name).await {
+            Ok(_) => Ok(Ok(())),
+            Err(err) => {
+                error!("Undeployment failed: {err}");
+                Ok(Err(format!("Undeployment failed: {err}")))
+            }
+        }
+    }
+
+    #[instrument(level = "debug", skip(self, ctx), fields(model = %model))]
+    async fn put_model(
+        &self,
+        ctx: Option<Context>,
+        model: String,
+        lattice: Option<String>,
+    ) -> anyhow::Result<Result<(String, String), String>> {
+        let client = self.get_client(ctx).await?;
+        match client.put_manifest(&model).await {
+            Ok(response) => Ok(Ok(response)),
+            Err(err) => {
+                error!("Failed to store model: {err}");
+                Ok(Err(format!("Failed to store model: {err}")))
+            }
+        }
+    }
+
+    #[instrument(level = "debug", skip(self, ctx), fields(manifest = ?manifest))]
+    async fn put_manifest(
+        &self,
+        ctx: Option<Context>,
+        manifest: OamManifest,
+        lattice: Option<String>,
+    ) -> anyhow::Result<Result<(String, String), String>> {
+        let client = self.get_client(ctx).await?;
+
+        // Serialize the OamManifest into bytes
+        let manifest_bytes =
+            serde_json::to_vec(&manifest).context("Failed to serialize OAM manifest")?;
+
+        // Convert the bytes into a string
+        let manifest_string = String::from_utf8(manifest_bytes)
```

Comment:
I don't think you need to convert it into a string. The `ManifestLoader` trait is [implemented for `Vec<u8>`](https://docs.rs/wadm-client/latest/wadm_client/loader/trait.ManifestLoader.html#impl-ManifestLoader-for-Vec%3Cu8%3E). 

It might be worth putting an extra implementation for `OamManifest` when the proper features are enabled but that isn't necessary here